src/scanner.c

```diff
@@ -2122,29 +2122,28 @@ void GetChar ( void )
   GET_CHAR();
 
   /* handle escape equences                                              */
-  if ( *TLS(In) == '\\' ) {
-    GET_CHAR();
-    TLS(Value)[0] = GetEscapedChar();
-  }
-  else if ( *TLS(In) == '\n' ) {
-    SyntaxError("Newline not allowed in character literal");
-  }
-  /* put normal chars into 'TLS(Value)'                                  */
-  else {
-    TLS(Value)[0] = *TLS(In);
-  }
-
-  /* read the next character                                             */
-  GET_CHAR();
-
-  /* check for terminating single quote                                  */
-  if ( *TLS(In) != '\'' )
-    SyntaxError("Missing single quote in character constant");
+  if ( *TLS(In) == '\n' ) {
+    SyntaxError("Character literal must not include <newline>");
+  } else {
+    if ( *TLS(In) == '\\' ) {
+      GET_CHAR();
+      TLS(Value)[0] = GetEscapedChar();
+    } else {
+      /* put normal chars into 'TLS(Value)' */
+      TLS(Value)[0] = *TLS(In);
+    }
+    TLS(Symbol) = S_CHAR;
```

Comment:
`GET_CHAR` (essentially...) advances the pointer into the input stream buffer; here we are looking to get the closing `'`.

This refactor is an attempt at  making the function less obscure, but maybe I am not very successful at it yet. 

The `TLS(Symbol) = S_CHAR` should probably happen at the end of the function, because only then can we be sure that we (successfully) parsed a character literal.